js2-mode.el

```diff
@@ -10878,6 +10878,69 @@ And, if CHECK-ACTIVATION-P is non-nil, use the value of TOKEN."
         (js2-check-activation-name s (or token js2-NAME)))
     name))
 
+;;; Use AST to extract semantic information
+
+(defun js2-get-element-index-from-array-node (elem array-node &optional hardcoded-array-index)
+  "Get index of ELEM from ARRAY-NODE or 0 and return it as string."
+  (let ((idx 0) elems (rlt hardcoded-array-index))
+    (setq elems (js2-array-node-elems array-node))
+    (if (and elem (not hardcoded-array-index))
+        (setq rlt (catch 'nth-elt
+                    (dolist (x elems)
+                      ;; Is same node by comparing position
+                      ;; We already known the ELEM does belong to ARRAY-NODE
+                      ;; so avoid comparing node type for performance reason
+                      (if (= (aref elem 2) (aref x 2))
+                        (throw 'nth-elt idx))
+                      (setq idx  (1+ idx)))
+                    0)))
+    (format "[%s]" rlt)))
+
+(defun js2-print-json-path (&optional hardcoded-array-index)
+  "Print the path to the JSON value under point, and save it in the kill ring.
+If HARDCODED-ARRAY-INDEX provided, actual array index in JSON path are replaced with it"
+  (interactive "P")
+  (let (previous-node current-node
+        key-name
+        (i 0) (json-path-max-depth 128)
+        rlt)
+
+    ;; The `js2-node-at-point' starts scanning from AST root node.
+    ;; So there is no way to optimize it.
+    (setq current-node (js2-node-at-point))
+
+    (while (and current-node
```

Comment:
`js2-node-at-point` always returns a non-nil value, and it's always a node. Why check that it's an array, or its length?